### ISSUE TITLE
Correctly specifying max_steps when accumulate_grad_batches > 1 

### ISSUE THREAD

**Issue by avasbr** [Lightning-AI/pytorch-lightning#19684]
### Bug description

Apologies if this has come up before, but I was not able to find the documentation I need. I'm seeing some odd behavior when specifying `max_steps` and `accumulate_grad_batches > 1`. I prefer to work in terms of `steps` rather than `epochs`, but for the sake of argument, let's assume that I pre-specify the # of epochs, but still want to use `max_steps`. I calculate max steps in the following way:

```
num_steps_per_epoch = int(num_train_samples // global_batch_size)  # using drop_last=True in the train dataloader
max_steps = num_epochs * num_steps_per_epoch
```
Everything works smoothly but as soon as I increase `accumulate_grad_batches` beyond 1, training stops prematurely. If i specify `max_epochs` directly vs using `max_steps` however, it'll always be correct, so perhaps it could be how I compute `max_steps` above, but I don't believe this to be the case.. if `max_steps` depends on the `global_step`, i.e. calls to `optimizer.step()`, then the number of accumulated batches shouldn't factor in. Am I missing something? see boring model for repro


### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
import torch
from torch.utils.data import DataLoader, Dataset
import pytorch_lightning as pl
from pytorch_lightning.callbacks import LearningRateMonitor
import torch
from pytorch_lightning import LightningModule


class RandomDataset(Dataset):
    def __init__(self, size: int, num_samples: int) -> None:
        """
        Args:
            size: tuple
            num_samples: number of samples
        """
        if num_samples < 1:
            raise ValueError("Provide an argument greater than 0 for `num_samples`")

        self.len = num_samples
        self.data = torch.randn(num_samples, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self) -> int:
        return self.len


# a few constants to match the values of the original problem
num_train_samples = 2500
num_val_samples = 797
num_grad_accumulate = 8
global_batch_size = 512
num_epochs = 10
num_gpus = 1

num_feats = 1024


# N=batch size, K = # of grad accumulations;
# for DDP, each GPU  gets N*K examples, so divide by K to get equivalent behavior for when K=1
# https://lightning.ai/docs/pytorch/stable/advanced/training_tricks.html
batch_size_per_gpu = int(global_batch_size / (num_grad_accumulate * num_gpus))
max_steps = num_epochs * int(num_train_samples // global_batch_size)

# run validation after every epoch

# counter-intuitive behavior where val_check_interval is based on # of forward passes
# rather than # of global steps (i.e. optimization steps). this means that we have to keep track
# of how many forward passes are occurring, which involves num_grad_accumulate.
# https://github.com/Lightning-AI/pytorch-lightning/issues/12205
num_forward_steps_per_epoch = int(
    num_train_samples // (global_batch_size / num_grad_accumulate)
)  # drop_last = True, hence the use of //
val_check_interval = num_forward_steps_per_epoch


class BoringData(pl.LightningDataModule):
    def __init__(self):
        super().__init__()

    def setup(self, stage=None):
        self.train_ds = RandomDataset(num_feats, num_train_samples)
        self.val_ds = RandomDataset(num_feats, num_val_samples)

    def train_dataloader(self):
        return DataLoader(
            self.train_ds,
            num_workers=8,
            batch_size=batch_size_per_gpu,
            shuffle=True,
            drop_last=True,
            pin_memory=True,
        )

    def val_dataloader(self):
        return DataLoader(
            self.val_ds,
            num_workers=8,
            batch_size=128,  # hardcoded for now
            shuffle=False,
            drop_last=False,
            pin_memory=True,
        )


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.model = torch.nn.Sequential(
            torch.nn.Linear(num_feats, int(num_feats // 2)),
            torch.nn.Linear(int(num_feats // 2), 2),
        )

    def forward(self, x):
        return self.model(x)

    def loss(self, batch, prediction):
        # An arbitrary loss to have a loss that updates the model weights during `Trainer.fit` calls
        return torch.nn.functional.mse_loss(prediction, torch.ones_like(prediction))

    def training_step(self, batch, batch_idx):
        output = self.model(batch)
        loss = self.loss(batch, output)
        self.log(
            "train/loss",
            loss,
            on_epoch=True,
            prog_bar=True,
            enable_graph=True,
            logger=True,
        )
        return loss

    def validation_step(self, batch, batch_idx):
        output = self.model(batch)
        loss = self.loss(batch, output)
        self.log("val/loss", loss, on_epoch=True, prog_bar=True)
        return loss

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.model.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]


model = BoringModel()
data = BoringData()
lr_monitor = LearningRateMonitor(logging_interval="step")

# Initialize a trainer
trainer = pl.Trainer(
    accumulate_grad_batches=num_grad_accumulate,
    devices=num_gpus,
    max_steps=max_steps,
    val_check_interval=val_check_interval,
    log_every_n_steps=1,
    num_sanity_val_steps=0,
    precision="16-mixed",
    callbacks=[lr_monitor],
    accelerator="cuda",
    strategy="auto",
)
trainer.fit(model, data)
```


### Error messages and logs

```
Using 16bit Automatic Mixed Precision (AMP)
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3]

  | Name  | Type       | Params
-------------------------------------
0 | model | Sequential | 525 K 
-------------------------------------
525 K     Trainable params
0         Non-trainable params
525 K     Total params
2.103     Total estimated model params size (MB)
Epoch 7: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 39/39 [00:00<00:00, 51.52it/s, v_num=49, train/loss_step=0.152, val/loss=0.253, train/loss_epoch=0.184]`Trainer.fit` stopped: `max_steps=40` reached.                                                                                                                                                                                                                                                                                                                             
Epoch 7: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 39/39 [00:00<00:00, 51.40it/s, v_num=49, train/loss_step=0.152, val/loss=0.253, train/loss_epoch=0.184]
```

I'm not sure why this stopped after only 8 epochs instead of 10. If I use `accumulate_grad_batches = 1`, or use `max_epochs = 10` instead of `max_steps`, everything works as intended. 

### Environment

PyTorch Lightning Version 2.0.0



### More info

_No response_

**Comment by avasbr** [comment]
Answering my own question and closing. Turns out the tricky issue is with using `drop_last=True` alongside `accumulate_grad_batch > 1` ; since the final batch in an epoch does not contribute towards the accumulated gradient, we wait until the next epoch to make the step and it throws off the counts. In this scenario, it might just make sense to rely on `max_epochs` instead or to use `drop_last=False`, in which case the math works out again... though that has its own consequences if the final batch is extremely small.
